server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -285,27 +285,54 @@ public TemporalAccessor parse(String input) {
      * @throws DateTimeParseException when unable to parse with any parsers
      */
     private TemporalAccessor doParse(String input) {
-        if (parsers.size() > 1) {
-            Object object = null;
-            if (canCacheLastParsedFormatter && lastParsedformatter != null) {
-                ParsePosition pos = new ParsePosition(0);
-                object = lastParsedformatter.toFormat().parseObject(input, pos);
+        if (parsers.size() != 1) {
+            // Copy to local from volatile to avoid error in multi-threads,
```

Comment:
In this particular case I don't see any issues with other threads updating `lastParsedformatter`: the only problem could be `null` value but this field is never reset to `null `

Reply:
There is optimization I'm trying to do here, in old impl, if there are more than 1 parsers, let's say there are N parsers for N formats, if the input value is not a string in any of the date formats, old impl will try to parse N + 1 times, hence in the for loop after attempting to parse with lastParsedformatter I added an condition check e.g. formatter != lastParsedformatter to avoid parsing the string values using the lastParsedformatter again. This should generally improve the performance in data-time  detection because most of the fields in a documents are not data-time, parsing the data-time is expensive especially when consider this method is invoked billions of times every day(our case, we have multi billions documents in one cluster. 

Given optimization I mentioned above, the lastParsedformatter will be read N + 1 times, in one call , if there is other threads update lastParsedformatter(which is very likely to happen), it may get inconsistent value of lastParsedformatter during one method call. 

Also, reading N+1 times from local cachedFormatter should faster than reading from N+1 times from lastParsedformatter since lastParsedformatter is volatile.